backend/wmg/api/rollup.py

```diff
@@ -108,7 +143,38 @@ def rollup_across_cell_type_descendants(df, cell_type_col="cell_type_ontology_te
     array_to_sum[tuple(dim_indices)] = numeric_df.to_numpy()
 
     cell_types = cell_type_column.unique()
-    descendants = find_descendants_per_cell_type(cell_types)
+
+    summed = rollup_across_cell_type_descendants_array(array_to_sum, cell_types)
+
+    # extract numeric data and write back into the dataframe
+    summed = summed[tuple(dim_indices)]
+    dtypes = numeric_df.dtypes
+    for col, array in zip(numeric_df.columns, summed.T):
+        df[col] = array.astype(dtypes[col])
+
+    return df
+
+
+def rollup_across_cell_type_descendants_array(array_to_sum, cell_types) -> np.ndarray:
```

Comment:
The logic here was factored out into a separate function so that the marker gene calculation code can perform data roll-up on a numpy array directly (instead of needing to reshape it into a tidy dataframe).